iot-hub/Samples/device/DeviceReconnectionSample/DeviceReconnectionSample.cs

```diff
@@ -0,0 +1,225 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Azure.Devices.Client.Exceptions;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client.Samples
+{
+    public class DeviceReconnectionSample
+    {
+        private static readonly Random s_randomGenerator = new Random();
+        private const int TemperatureThreshold = 30;
+        private static readonly TimeSpan s_sleepDuration = TimeSpan.FromSeconds(5);
+        private static readonly TimeSpan s_operationTimeout = TimeSpan.FromHours(1);
+
+        private readonly string _deviceConnectionString;
+        private readonly TransportType _transportType;
+        private readonly ILogger _logger;
+        private readonly object _lock = new object();
+
+        private static DeviceClient s_deviceClient;
+        private static ConnectionStatus s_connectionStatus;
+        private static bool s_wasEverConnected;
+
+        public DeviceReconnectionSample(string deviceConnectionString, TransportType transportType, ILogger logger)
+        {
+            _logger = logger;
+            _deviceConnectionString = deviceConnectionString;
+            _transportType = transportType;
+
+            InitializeClient();
+        }
+
+        private void ConnectionStatusChangeHandler(ConnectionStatus status, ConnectionStatusChangeReason reason)
+        {
+            _logger.LogDebug($"Connection status changed: status={status}, reason={reason}");
+
+            s_connectionStatus = status;
+            switch (s_connectionStatus)
+            {
+                case ConnectionStatus.Connected:
+                    _logger.LogDebug("### The DeviceClient is CONNECTED; all operations will be carried out as normal.");
+
+                    s_wasEverConnected = true;
+                    break;
+
+                case ConnectionStatus.Disconnected_Retrying:
+                    _logger.LogDebug("### The DeviceClient is retrying based on the retry policy. Do NOT close or open the DeviceClient instance");
+                    break;
+
+                case ConnectionStatus.Disabled:
+                    _logger.LogDebug("### The DeviceClient has been closed gracefully." +
+                        "\nIf you want to perform more operations on the device client, you should dispose (DisposeAsync()) and then open (OpenAsync()) the client.");
+                    break;
+
+                case ConnectionStatus.Disconnected:
+                    {
```

Comment:
Curious the reason for the extra braces

Reply:
yup, doesn't need the extra braces there, will remove